### PR TITLE
[ABI-Break][SYCL] Fix ext_oneapi_cl_profile to be ABI-neutral

### PR DESCRIPTION
**Pull request by bso-intel**
Returning a std::string from device.cpp to device.hpp will cross the ABI boundary.
So, we need to neutralize it via detail::string.

### CODE REVIEW

**Comment by bso-intel** [comment]
@stdale-intel 
This PR is breaking ABI.
Could you give me an exception for breaking ABI since this is an urgently requested fix for the customers?
Thanks.

File: sycl/source/device.cpp
Comment (againull) [inline sycl/source/device.cpp:317]: Unfortunately this introduces unspecified `device::ext_oneapi_cl_profile_impl ` method to the `sycl::device` class, we don't usually expose such implementation methods in sycl classes.

Please hide this implementation method into device_impl class. Please see, for example, ext_oneapi_can_compile implentation in this file:
// kernel_compiler extension methods
```
bool device::ext_oneapi_can_compile(
    ext::oneapi::experimental::source_language Language) {
  return impl->extOneapiCanCompile(Language);
}
```

**Review by againull — CHANGES_REQUESTED** [review]

File: sycl/source/device.cpp
Comment (bso-intel) [reply sycl/source/device.cpp:317]: Actually, this method belongs to the private section, so this change is not publicize unspecified method to the class.
Users won't be able to use this method like all other private methods in the `sycl::device` class.

Comment (againull) [reply sycl/source/device.cpp:317]: I see, technically that works, and I see other private methods there but they were all added in scope of same abi effort, all other implementation methods are in device_impl - this is our regular approach. adding method to device_imple instead of private section also means you don't need to introduce new symbol into linux/windows dumps.
 
is there any reason why we use private methods approach just for abi related patches?

Comment (againull) [reply sycl/source/device.cpp:317]: Ok, let me approve considering that similar private methods has been already added for the same reason.

**Review by againull — APPROVED** [review]

**Review by stdale-intel — APPROVED** [review]

**Review by stdale-intel — APPROVED** [review]
changes are approved outside window because this is required for pytorch upstreaming and will be cherr ypicked into other branchs where its allowed

**Comment by bso-intel** [comment]
@intel/llvm-gatekeepers 
Please merge this PR as soon as possible once all tests pass.
This PR is urgently requested by our customer.
Thanks.

File: sycl/source/device.cpp
Comment (AlexeySachkov) [reply sycl/source/device.cpp:317]: > is there any reason why we use private methods approach just for abi related patches?

My understanding is that we can't change signature of a public method. Even if we had (or already have) some implicit conversions from `detail::string` to `std::string`, returning the former from the public method could still break someone's code. For example because they are storing the result into `their::string` also using implicit conversion from `std::string` being completely unaware of our `detail::string`.

`impl` is already within the library so even if we add `impl->right_method` there will still be `std::string` crossing ABI boundary that is part of the public API.

File: sycl/test/abi/sycl_symbols_windows.dump
Comment (bso-intel) [reply sycl/test/abi/sycl_symbols_windows.dump:3773]: The script automatically generates different number of symbols.
The number of symbols in linux dump is 3988 and the number of symbols in windows dump is 4282.
So, there is no 1-1 relationship between two files.

**Comment by stdale-intel** [comment]
@intel/llvm-gatekeepers  can we pls merge this? All test pass but AMD and this is not platform specific feature, so no reason to believe it will fail on AMD. This is holding up downstream efforts

**Comment by bader** [comment]
@stdale-intel, are ABI breaks still allowed? The [documentation](https://github.com/intel/llvm/blob/sycl/sycl/doc/developer/ABIPolicyGuide.md#abi-breaking-changes-window-april-18---july-11-2023-closed) hasn't been updated for the recent period, so it's not clear if this change can be accepted.

If yes, can we merge other PRs with `abi-break` [label](https://github.com/intel/llvm/labels/abi-break)?

I would appreciate, we explicitly document the rules for making such decisions.

**Comment by stdale-intel** [comment]
@bader  this is a special one off case as required for the larger pytorch upstream effort. No other/general ABI breaks should be allowed/considered.

**Comment by bader** [comment]
> All test pass but AMD and this is not platform specific feature, so no reason to believe it will fail on AMD. This is holding up downstream efforts

@bso-intel, @stdale-intel, please, keep an eye on pre-commit results and revert/fix ASAP if tests fail on AMD.

**Comment by AlexeySachkov** [comment]
> @stdale-intel, are ABI breaks still allowed? The [documentation](https://github.com/intel/llvm/blob/sycl/sycl/doc/developer/ABIPolicyGuide.md#abi-breaking-changes-window-april-18---july-11-2023-closed) hasn't been updated for the recent period, so it's not clear if this change can be accepted.
> 
> If yes, can we merge other PRs with `abi-break` [label](https://github.com/intel/llvm/labels/abi-break)?
> 
> I would appreciate, we explicitly document the rules for making such decisions.

I would also highlight that this particular patch is about `experimental` extension and my understanding is that we are allowed to break their ABI even outside the window

**Pull request merged**
